bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -551,16 +551,27 @@ private static boolean isConvertibleToArrayType(Object sourceValue, BArrayType t
         }
         ArrayValue source = (ArrayValue) sourceValue;
         Type targetTypeElementType = targetType.getElementType();
-        if (source.getType().getTag() == TypeTags.ARRAY_TAG) {
-            Type sourceElementType = ((BArrayType) source.getType()).getElementType();
+        Type sourceType = source.getType();
+        if (sourceType.getTag() == TypeTags.ARRAY_TAG) {
+            Type sourceElementType = ((BArrayType) sourceType).getElementType();
             if (isNumericType(sourceElementType) && isNumericType(targetTypeElementType)) {
                 return true;
             }
         }
+        int targetSize = targetType.getSize();
+        long sourceSize = source.getLength();
+        boolean returnVal = true;
+        if (!TypeChecker.hasFillerValue(sourceType) && sourceSize < targetSize) {
+            addErrorMessage(0, errors, "array size '" + targetSize + "' is out of range for source type'" +
```

Comment:
Shall we improve the error that source type does not have filler values as well.